content/blog/2017/2017-01-18-refactoring-conditional-to-pipeline.adoc

```diff
@@ -0,0 +1,475 @@
+
+image::/images/post-images/2017-01-17/freestyle-conditional-config.png[Freestyle Job Conditional BuildStep, role="center", width=800]
+
+Here's the output when I run this project with `REQUESTED_ACTION` set to "greeting":
+
+[source]
+----
+Run condition [Strings match] enabling prebuild for step [Execute shell]
+Strings match run condition: string 1=[greeting], string 2=[greeting]
+Run condition [Strings match] enabling perform for step [Execute shell]
+[freestyle-conditional] $ /bin/sh -xe /var/folders/hp/f7yc_mwj2tq1hmbv_5n10v2c0000gn/T/hudson5963233933358491209.sh
++ echo 'Hello, bitwiseman!'
+Hello, bitwiseman!
+Finished: SUCCESS
+----
+
+And when I pass the value "silence":
+
+[source]
+----
+Run condition [Strings match] enabling prebuild for step [Execute shell]
+Strings match run condition: string 1=[greeting], string 2=[greeting]
+Run condition [Strings match] enabling perform for step [Execute shell]
+[freestyle-conditional] $ /bin/sh -xe /var/folders/hp/f7yc_mwj2tq1hmbv_5n10v2c0000gn/T/hudson5963233933358491209.sh
++ echo 'Hello, bitwiseman!'
+Hello, bitwiseman!
+Finished: SUCCESS
+----
+
+This is a simple example but the conditional step can contain any regular build step.
+When combined with other plugins, it can control whether to send notifications,
+gather data from other sources, wait for user feedback, or call other projects.
+
+The Conditional BuildStep plugin does a great job of leveraging strengths of
+the Jenkins web UI, Freestyle jobs, and UI-based programming,
+but it is also hampered by their limitations.
+The Jenkins web UI can be clunky and confusing at times.
+Like the steps in any freestyle job, these conditional steps are only
+stored and viewable in Jenkins.
+They are not versioned with other product or build code and can't be code reviewed.
+Like any number of UI-based programming tools, it has to make trade-offs between clarity
+and flexibility: more options or clearer presentation.
+There's only so much space on the screen.
+
+== Refactoring to Pipeline
+
+Jenkin Pipeline, on the other hand, enables users to implement a pipeline as code.
+Pipeline code can be written directly in the Jenkins Web UI or in any text editor.
+It is a full-featured programming language,
+which gives users access to much broader set of conditional statements
+without the restrictions of UI-based programming.
+
+So, taking the example above, the Pipeline equivalent is:
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+    agent any
+    parameters {
+        choice(
+            // choices must be newline separated
+            choices: ['greeting','silence'].join('\n'),
+            description: '',
+            name: 'REQUESTED_ACTION')
+    }
+
+    stages {
+        stage ('Speak') {
+            when {
+                // Only say hello if a "greeting" is requested
+                expression { return params.REQUESTED_ACTION == 'greeting' }
```

Comment:
Minor: I would guess that the `return` keyword isn't necessary, but I don't know whether Declarative supports that Groovy syntax.